vhost-device-gpu/src/device.rs

```diff
@@ -520,75 +559,78 @@ impl VhostUserGpuBackendInner {
                 }
             }
             POLL_EVENT => {
                 trace!("Handling POLL_EVENT");
-                virtio_gpu.event_poll();
+                renderer.event_poll();
             }
             _ => {
-                warn!("unhandled device_event: {}", device_event);
+                warn!("unhandled device_event: {device_event}");
                 return Err(Error::HandleEventUnknown.into());
             }
         }
 
         Ok(())
     }
 
+    fn extract_backend_and_vring<'a>(
+        &mut self,
+        vrings: &'a [VringRwLock],
+    ) -> IoResult<(&'a VringRwLock, GpuBackend)> {
+        let control_vring = &vrings[CONTROL_QUEUE as usize];
+        let backend = self
+            .gpu_backend
+            .take()
+            .ok_or_else(|| io::Error::other("set_gpu_socket() not called, GpuBackend missing"))?;
+        Ok((control_vring, backend))
+    }
+
     fn lazy_init_and_handle_event(
         &mut self,
         device_event: u16,
         evset: EventSet,
         vrings: &[VringRwLock],
         _thread_id: usize,
     ) -> IoResult<Option<EventFd>> {
-        // We use thread_local here because it is the easiest way to handle VirtioGpu
-        // being !Send
-        thread_local! {
-            static VIRTIO_GPU_REF: RefCell<Option<RutabagaVirtioGpu>> = const { RefCell::new(None) };
-        }
-
         debug!("Handle event called");
         if evset != EventSet::IN {
             return Err(Error::HandleEventNotEpollIn.into());
-        };
-
-        let mut event_poll_fd = None;
-        VIRTIO_GPU_REF.with_borrow_mut(|maybe_virtio_gpu| {
-            let virtio_gpu = match maybe_virtio_gpu {
-                Some(virtio_gpu) => virtio_gpu,
-                None => {
-                    let gpu_backend = self.gpu_backend.take().ok_or_else(|| {
-                        io::Error::other("set_gpu_socket() not called, GpuBackend missing")
-                    })?;
-
-                    // We currently pass the CONTROL_QUEUE vring to RutabagaVirtioGpu, because we
-                    // only expect to process fences for that queue.
-                    let control_vring = &vrings[CONTROL_QUEUE as usize];
-
-                    // VirtioGpu::new can be called once per process (otherwise it panics),
-                    // so if somehow another thread accidentally wants to create another gpu here,
-                    // it will panic anyway
-                    let virtio_gpu =
-                        RutabagaVirtioGpu::new(control_vring, &self.gpu_config, gpu_backend);
-                    event_poll_fd = virtio_gpu.get_event_poll_fd();
-
-                    maybe_virtio_gpu.insert(virtio_gpu)
-                }
-            };
-
-            self.handle_event(device_event, virtio_gpu, vrings)
-        })?;
+        }
 
-        Ok(event_poll_fd)
+        match self.gpu_config.gpu_mode() {
+            #[cfg(feature = "backend-gfxstream")]
```

Comment:
Will this match fail to compile if no features are enabled? (Can't check myself at the moment)

Reply:
@dorindabassey okay, but what about using `--no-default-features` ?

Reply:
right, doesn't make sense. we don't have a dummy backend like null. I think it's better we don't have null because GPU backend is a core functionality. For sound "null" make sense for some use cases, but for GPU no backend is useless.